code/datums/components/nonwall_pierce.dm

```diff
@@ -0,0 +1,24 @@
+/datum/component/nonwall_pierce
+
+TYPEINFO(/datum/component/nonwall_pierce)
+	initialization_args = list()
+
+/datum/component/nonwall_pierce/Initialize()
+	. = ..()
+	if(!istype(parent, /obj/projectile))
+		return COMPONENT_INCOMPATIBLE
+	RegisterSignal(parent, COMSIG_OBJ_PROJ_COLLIDE, PROC_REF(update_pierces))
+
+
+/datum/component/nonwall_pierce/proc/update_pierces(var/obj/projectile/P, var/atom/hit)
+	var/turf/T = get_turf(hit)
+	if (isrestrictedz(T.z))
+		return 0
+	if (istype(hit, /obj/machinery/door))
+		var/obj/machinery/door/door = hit
+		return !door.density
+	return PROJ_ATOM_PASSTHROUGH * (!iswall(T) || (istype(hit, /obj/machinery/door)))
```

Comment:
Condition here feels weird. It won't pierce things if there's a wall on the turf (hitting the thing instead of the wall?)
And the door handling in the return line feels off

Reply:
ooer. you're right. lemme cover that